saya/core/src/block_ingestor/polling.rs

```diff
@@ -1,127 +1,275 @@
-                );
+        match block_number {
+            Ok(block_number) => Some(block_number),
+            Err(err) => {
+                error!("Failed to fetch latest block: {}", err);
+                None
+            }
+        }
+    }
 
-                tokio::select! {
-                    _ = self.finish_handle.shutdown_requested() => break,
-                    _ = tokio::time::sleep(PROVE_BLOCK_FAILURE_BACKOFF) => continue,
-                }
+    /// Worker function: proves a block and sends the result.
+    async fn worker(
+        task_rx: Arc<Mutex<mpsc::Receiver<u64>>>,
+        block_pie_generator: B,
+        finish_handle: FinishHandle,
+        rpc_url: Url,
+        channel: mpsc::Sender<BlockInfo>,
+        snos: S,
+        db: DB,
+    ) where
+        S: AsRef<[u8]> + Send + Sync + 'static,
+        B: BlockPieGenerator + Send + Sync + 'static,
+        DB: PersistantStorage + Send + Sync + 'static,
+    {
+        loop {
+            let block_number = if let Some(block_number) = task_rx.lock().await.recv().await {
+                block_number
+            } else {
+                break;
+            };
+
+            if finish_handle.is_shutdown_requested() {
+                break;
             }
 
-            let pie = match prove_block::prove_block(
-                self.snos.as_ref(),
-                self.current_block,
-                // This is because `snos` expects a base URL to be able to derive `pathfinder` RPC path.
-                self.rpc_url.as_str().trim_end_matches("/rpc/v0_7"),
-                LayoutName::all_cairo,
-                true,
+            crate::utils::retry_with_backoff(
+                || db.initialize_block(block_number.try_into().unwrap()),
+                "initialize_block",
+                MAX_RETRIES as u32,
+                Duration::from_secs(5),
             )
             .await
-            // Need to do this as `ProveBlockError::ReExecutionError` is not `Send`
-            .map_err(|err| format!("{}", err))
-            {
-                Ok((pie, _)) => pie,
-                Err(err) => {
-                    error!("Failed to prove block #{}: {}", self.current_block, err);
+            .unwrap();
 
-                    tokio::select! {
-                        _ = self.finish_handle.shutdown_requested() => break,
-                        _ = tokio::time::sleep(PROVE_BLOCK_FAILURE_BACKOFF) => continue,
+            match db
+                .get_pie(block_number.try_into().unwrap(), Step::Snos)
+                .await
+            {
+                Ok(pie_bytes) => match CairoPie::from_bytes(&pie_bytes) {
+                    Ok(_pie) => {
+                        let new_block = BlockInfo {
+                            number: block_number,
+                            status: BlockStatus::SnosPieGenerated,
+                        };
+                        log::trace!("Pie generated for block #{}", block_number);
```

Comment:
Not for this PR specifically, but we should move to structured logging instead of formatting, since it will be beneficial for monitoring later.

```rust
log::trace!("Pie generated for block #{}", block_number);
```
would be
```rust
log::trace!(block_number, "Generated PIE.");
```